# ECS0005: Prefer FormattableString for culture-specific strings

This rule is discussed in detail in [Effective C#: 50 Specific Ways to Improve your C#](https://www.oreilly.com/library/view/effective-c-50/9780134579290/). Guidance about the feature can be found on the [.NET Blog - String Interpolation in C# 10 and .NET 6](https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/).

## Cause

The rule is triggered when a `string` is used for an interpolated string that could benefit from culture-specific formatting. 

## Rule description

There are still valid use cases for `string.Format` and `StringBuilder.AppendFormat` where the composite format string isn't known at compile time (e.g., localized resources). However, if the string would otherwise be hardcoded into the C#, interpolated are preferred (from a performance perspective). Using `FormattableString` or `string.Create` instead of `string` for interpolated strings ensures that culture-specific formatting is correctly applied, preventing issues where the default culture may lead to incorrect string representations.

## How to fix violations

Replace the `string` with `FormattableString` where culture-specific formatting is required. For example, use `FormattableString.Invariant(...)` or `FormattableString.ToString(IFormatProvider)` as needed.

## When to suppress warnings

Suppress warnings if the default culture is explicitly intended, or the string does not require culture-specific formatting.

### Suppress a warning

If you want to suppress a single violation, add preprocessor directives to your source file to disable and then re-enable the rule.

```csharp
#pragma warning disable ECS0005
// The code that's violating the rule
#pragma warning restore ECS0005
```

To disable the rule for a file, folder, or project, set its severity to none in the [configuration file](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files).

```ini
[*.cs]
dotnet_diagnostic.ECS0005.severity = none
```

## Example of a violation

### Description

Using `string` for an interpolated string that requires culture-specific formatting.

### Code

```csharp
public string GetMessage()
{
    double value = 299792.458;
    return $"The speed of light is {value:N3} km/s."; // ECS0005 triggers here
}
```

## Example of how to fix

### Description

Replace `string` with `FormattableString` to handle culture-specific formatting.

### Code

```csharp
public string GetMessage()
{
    double value = 299792.458;
    return FormattableString.Invariant($"The speed of light is {value:N3} km/s.");
}
```

## Related rules

[ECS0004: Replace string.Format with interpolated string](./ECS0004.md)